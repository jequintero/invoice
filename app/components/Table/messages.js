/*
 * Table Messages
 *
 * This contains all the text for the Table component.
 */

import { defineMessages } from 'react-intl';

export const scope = 'app.components.Table';

export default defineMessages({
  header: {
    id: `${scope}.header`,
    defaultMessage: 'This is the Table component!',
  },
  noItems: {
    id: `${scope}.noItems`,
    defaultMessage: 'No items to show',
  },
  itemName: {
    id: `${scope}.noItems`,
    defaultMessage: 'Item name',
  },
  quantity: {
    id: `${scope}.noItems`,
    defaultMessage: 'Quantity',
  },
  price: {
    id: `${scope}.noItems`,
    defaultMessage: 'Price',
  },
});
