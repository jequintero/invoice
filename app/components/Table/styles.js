import styled from 'styled-components';

export const TableContainer = styled.table`
  width: 100%;
  color: #b1b4c7;
  font-family: 'Open Sans';
  text-align: left;
  border-collapse: separate;
  border-spacing: 0 5px;
`;

export const TableHeader = styled.th`
  letter-spacing: 2px;
  text-transform: uppercase;
  font-weight: 300;
  padding: 20px;
`;

export const TableRow = styled.tr`
  background-color: #fff;
  color: #303863;
  box-shadow: 0 3px 8px #cccccc6e;
  transition: all 0.3s ease;
  td:first-child {
    border-top-left-radius: 4px;
    border-bottom-left-radius: 4px;
  }
  td:last-child {
    border-top-right-radius: 4px;
    border-bottom-right-radius: 4px;
  }
  &:hover {
    transform: scale(1.02);
    box-shadow: 0 4px 20px #cccccc6e;
  }
  &:hover img {
    cursor: pointer;
    opacity: 1;
    transform: scale(0.9);
  }
  &.onFocus {
    transform: scale(1.02);
    background-color: #e9fff8;
    img {
      opacity: 1;
      transform: scale(0.9);
    }
  }
`;

export const TableData = styled.td`
  padding: 20px;
  position: relative;
`;

export const SumRow = styled.div`
  width: 30%;
  background-color: #fff;
  color: #303863;
  box-shadow: 0 3px 8px #cccccc6e;
  margin-bottom: 5px;
  border-radius: 4px;
  display: flex;
  flex-direction: column;
  padding: 20px;
`;

export const Sum = styled.div`
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  margin-top: 10px;
`;

export const SumValue = styled.div`
  width: 45%;
`;

export const Button = styled.img`
  position: absolute;
  right: 20px;
  width: 30px;
  height: 30px;
  opacity: 0;
  transform: scale(0.5);
  transition: all 0.3s ease;
`;

export const Input = styled.input`
  outline: none;
  cursor: pointer;
  min-width: 30px;
  &:focus {
    color: #969696;
  }
`;

export const Title = styled.p`
  color: #9699ad;
  font-weight: 600;
  text-align: center;
  text-transform: uppercase;
`;

export const Column = styled.div`
  display: flex;
  margin: 10px;
  justify-content: space-between;
`;
