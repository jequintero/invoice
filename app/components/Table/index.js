/**
 *
 * Table
 *
 */

import React from 'react';
import { sumBy, round } from 'lodash';
import PropTypes from 'prop-types';
import {
  TableContainer,
  TableHeader,
  TableRow,
  TableData,
  SumRow,
  Sum,
  SumValue,
  Button,
  Input,
  Title,
  Column,
} from './styles';
import cancel from './assets/cancel.svg';
import messages from './messages';
/* eslint-disable react/prefer-stateless-function */
const ROUND_VALUE = 2;
const TAX = 5;
class Table extends React.Component {
  deleteItem = event => {
    const { deleteItem } = this.props;
    deleteItem(event.target.id);
  };

  onFocus = event => {
    const { setFocus } = this.props;
    setFocus(event.currentTarget.dataset.index, true);
  };

  onBlur = event => {
    const { setFocus } = this.props;
    setFocus(event.currentTarget.dataset.index, false);
  };

  onChange = event => {
    const { onChange } = this.props;
    onChange(
      event.currentTarget.dataset.type,
      event.currentTarget.dataset.index,
      event.target.value,
    );
  };

  handleKeyPress = (event, dataType, index) => {
    if (event.key === 'Enter') {
      const dataTypes = ['item', 'qty', 'price'];
      const currentPosition = dataTypes.findIndex(k => k === dataType);
      let founded = this[`${dataTypes[currentPosition + 1]}_${index}`];
      if (!founded) founded = this[`${dataTypes[0]}_${index + 1}`];
      if (!founded) event.target.blur();
      if (founded) {
        founded.focus();
        founded.select();
      }
    }
  };
  /* eslint-disable */
  render() {
    const { tableHeaders, tableData } = this.props;
    return (
      <div>
        {tableData.length ? <TableContainer>
          <tbody>
            <tr>
              {tableHeaders.map(header => (
                <TableHeader key={header}>{header}</TableHeader>
              ))}
            </tr>
            {tableData.map((row, i) => (
              <TableRow key={i} className={row.focused ? 'onFocus' : ''}>
                <TableData>
                  <Input
                    onChange={this.onChange}
                    type="text"
                    autoFocus={row.focused ? 'autofocus' : false}
                    value={row.item}
                    onFocus={this.onFocus}
                    onBlur={this.onBlur}
                    data-type="item"
                    data-index={i}
                    placeholder={messages.itemName.defaultMessage}
                    ref={input => {
                      this[`item_${i}`] = input;
                    }}
                    onKeyPress={event => this.handleKeyPress(event, 'item', i)}
                  />
                </TableData>
                <TableData>
                  <Input
                    onChange={this.onChange}
                    type="number"
                    value={row.qty}
                    onFocus={this.onFocus}
                    onBlur={this.onBlur}
                    data-type="qty"
                    data-index={i}
                    placeholder={messages.quantity.defaultMessage}
                    ref={input => {
                      this[`qty_${i}`] = input;
                    }}
                    onKeyPress={event => this.handleKeyPress(event, 'qty', i)}
                  />
                </TableData>
                <TableData>
                  <Input
                    onChange={this.onChange}
                    type="number"
                    value={row.price}
                    onFocus={this.onFocus}
                    onBlur={this.onBlur}
                    data-type="price"
                    placeholder={messages.price.defaultMessage}
                    data-index={i}
                    ref={input => {
                      this[`price_${i}`] = input;
                    }}
                    onKeyPress={event => this.handleKeyPress(event, 'price', i)}
                  />
                </TableData>
                <TableData>
                  {`$${row.total}`}
                  <Button id={i} src={cancel} onClick={this.deleteItem} />
                </TableData>
              </TableRow>
            ))}
          </tbody>
        </TableContainer> : null}
        {!tableData.length && <Title>{messages.noItems.defaultMessage}</Title>}
        {tableData.length > 0 ?
          (<Sum>
            <SumRow>
              <Column>
                <SumValue>Subtotal</SumValue>
                <SumValue>{`$${round(sumBy(tableData, 'total'), ROUND_VALUE)}`}</SumValue>
              </Column>
              <Column>
                <SumValue>Tax (5%)</SumValue>
                <SumValue>{`$${round((sumBy(tableData, 'total') * TAX) / 100, ROUND_VALUE)}`}</SumValue>
              </Column>
              <Column>
                <SumValue>Total</SumValue>
                <SumValue>{`$${round((sumBy(tableData, 'total') + (sumBy(tableData, 'total') * TAX) / 100), ROUND_VALUE) }`}</SumValue>
              </Column>
            </SumRow>
          </Sum>) : (null)
        }
      </div>
    );
  }
}

Table.propTypes = {
  tableHeaders: PropTypes.array,
  tableData: PropTypes.array,
  deleteItem: PropTypes.func,
  setFocus: PropTypes.func,
  onChange: PropTypes.func,
};

Table.defaultProps = {
  tableHeaders: [],
  tableData: [],
};

export default Table;
