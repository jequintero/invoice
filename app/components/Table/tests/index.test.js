import React from 'react';
import { shallow } from 'enzyme';
// import { enzymeFind } from 'styled-components/test-utils';

import Table from '../index';

describe('<Table />', () => {
  it('Expect to have unit tests specified', () => {
    const wrapper = shallow(<Table />);
    expect(wrapper).toMatchSnapshot();
  });
});
