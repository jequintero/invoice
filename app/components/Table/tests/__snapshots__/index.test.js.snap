// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> Expect to have unit tests specified 1`] = `
<div>
  <styles__Title>
    No items to show
  </styles__Title>
</div>
`;
